### ISSUE TITLE
Mysmartblinds don't always react properly

### ISSUE THREAD

**Issue by rnewhart** [apexad/homebridge-mysmartblinds-bridge#64]
**Bug or issue information**
When I ask homekit to close mysmartblinds I get randomly on some shades:

[HOOBS 21EA@Target Position] The write handler for the characteristic 'Target Position' on the accessory 'Office Left' didn't respond at all!. Please check that you properly call the callback!

When I restarted the bridge I get this on all blinds
[HOOBS 21EA@Office Center@@target Position] characteristic was supplied illegal value: number -1 exceeded minimum of 0

I installed
Mysmartblinds Bridge
2.2.4 published 12/2/2020

There is no place to set target position in configuration

**Homebridge setup information**
HOOBS 3.3.3
> Include details about your configuration if needed, but do not include any email address or password.  
> Include node version, if known.

**Logs**
[HOOBS 21EA@Office Center@@Target Position] characteristic was supplied illegal value: number -1 exceeded minimum of 0
3/2/2021, 3:32:30 PM [HOOBS 21EA@Office Center@@Current Position] characteristic was supplied illegal value: number -1 exceeded minimum of 0
3/2/2021, 3:32:30 PM [HOOBS 21EA@Office Center@Office Center Battery@Battery Level] characteristic was supplied illegal value: number -1 exceeded minimum of 0
3/2/2021, 3:32:30 PM [HOOBS 21EA@Living Room Center@@Target Position] characteristic was supplied illegal value: number -1 exceeded minimum of 0
3/2/2021, 3:32:30 PM [HOOBS 21EA@Living Room Center@@Current Position] characteristic was supplied illegal value: number -1 exceeded minimum of 0
3/2/2021, 3:32:30 PM [HOOBS 21EA@Living Room Center@Living Room Center Battery@Battery Level] characteristic was supplied illegal value: number -1 exceeded minimum of 0
3/2/2021, 3:32:30 PM [HOOBS 21EA@Office Right@@Target Position] characteristic was supplied illegal value: number -1 exceeded minimum of 0
3/2/2021, 3:32:30 PM [HOOBS 21EA@Office Right@@Current Position] characteristic was supplied illegal value: number -1 exceeded minimum of 0
3/2/2021, 3:32:30 PM [HOOBS 21EA@Office Right@Office Right Battery@Battery Level] characteristic was supplied illegal value: number -1 exceeded minimum of 0
3/2/2021, 3:32:30 PM [HOOBS 21EA@Living Room Left@@Target Position] characteristic was supplied illegal value: number -1 exceeded minimum of 0
3/2/2021, 3:32:30 PM [HOOBS 21EA@Living Room Left@@Current Position] characteristic was supplied illegal value: number -1 exceeded minimum of 0
3/2/2021, 3:32:30 PM [HOOBS 21EA@Living Room Left@Living Room Left Battery@Battery Level] characteristic was supplied illegal value: number -1 exceeded minimum of 0

**Comment by apexad** [comment]
The API tends to give -1 when the bridge lost communication with the blinds. Try restarting the bridge.

**Comment by randymassey** [comment]
Hey there…. Randy touching base…. You know, of interest, I also began seeing this today too in my logging, but only after I upgraded HOOBS to 3.3.3.  I do not recall seeing this before,  & I THNK I noticed for the OP who opened this ticket, they have HOOBS 3.3.3 as well…

I have not had any operability issues tho.  Blinds / plugin are working per usual.

Alex, I wanted to (unrelated) but confirm something with you.  As I am running HOOBS, the default for node.js version is 12.19.0 (even with the update from them today), & I do see in the logs that your plugin would like to see 14.x.x (not recalling the precise rev top of brain).

Should I check in with HOOBS support to see if that needs to be updated?  I did try it once from articles online & it toasted HOOBS & I had to rebuild it, so I left that alone, since support for v 12.x.x is thru 2022, but I was having other issues at the time with the SleepIQ plugin which was ultimately disabled by the author due to a desist letter from Sleep Number & using the API, so I knew I was doing a rebuild anyway, so I ran an update to node.js, & it toasted HOOBS (I’ve since seen some warnings on the GitHub page read me, about doing precisely that).

I’ve observed this in the logs as feedback from the plugin having preference for 14.x.x for a while, but have never asked you about that.  I also wonder if that could be relating to the issues where I sometimes am getting the “device not responding” quip back from Siri on the HomePods when I tell it to open all blinds, even tho they usually do open, there’s just a delay I tend to notice?

Sorry I crossed topics in this message…

RM

**Comment by rnewhart** [comment]
I have rebooted the bridge with no help.  

3/2/2021, 9:47:33 PM [HOOBS 21EA@Target Position] The write handler for the characteristic 'Target Position' on the accessory 'Living Room Dining' was slow to respond!
3/2/2021, 9:47:33 PM [HOOBS 21EA@Target Position] The write handler for the characteristic 'Target Position' on the accessory 'Office Center' was slow to respond!
3/2/2021, 9:47:33 PM [HOOBS 21EA@Target Position] The write handler for the characteristic 'Target Position' on the accessory 'Living Room Center' was slow to respond!
3/2/2021, 9:47:33 PM [HOOBS 21EA@Target Position] The write handler for the characteristic 'Target Position' on the accessory 'Office Left' was slow to respond!
3/2/2021, 9:47:33 PM [HOOBS 21EA@Target Position] The write handler for the characteristic 'Target Position' on the accessory 'Living Room Right' was slow to respond!
3/2/2021, 9:47:33 PM [HOOBS 21EA@Target Position] The write handler for the characteristic 'Target Position' on the accessory 'Office Right' was slow to respond!
3/2/2021, 9:47:33 PM [HOOBS 21EA@Target Position] The write handler for the characteristic 'Target Position' on the accessory 'Living Room Left' was slow to respond!
3/2/2021, 9:47:36 PM [MySmartBlindsBridge] Office Left currentPosition is now 0
3/2/2021, 9:47:36 PM [MySmartBlindsBridge] Office Right currentPosition is now 0
3/2/2021, 9:47:39 PM [HOOBS 21EA@Target Position] The write handler for the characteristic 'Target Position' on the accessory 'Living Room Dining' didn't respond at all!. Please check that you properly call the callback!
3/2/2021, 9:47:39 PM [HOOBS 21EA@Target Position] The write handler for the characteristic 'Target Position' on the accessory 'Office Center' didn't respond at all!. Please check that you properly call the callback!
3/2/2021, 9:47:39 PM [HOOBS 21EA@Target Position] The write handler for the characteristic 'Target Position' on the accessory 'Living Room Center' didn't respond at all!. Please check that you properly call the callback!
3/2/2021, 9:47:39 PM [HOOBS 21EA@Target Position] The write handler for the characteristic 'Target Position' on the accessory 'Living Room Right' didn't respond at all!. Please check that you properly call the callback!
3/2/2021, 9:47:39 PM [HOOBS 21EA@Target Position] The write handler for the characteristic 'Target Position' on the accessory 'Living Room Left' didn't respond at all!. Please check that you properly call the callback!
3/2/2021, 9:47:45 PM [MySmartBlindsBridge] Office Center currentPosition is now 0
3/2/2021, 9:47:45 PM [MySmartBlindsBridge] Living Room Left currentPosition is now 0
3/2/2021, 9:47:45 PM [MySmartBlindsBridge] Living Room Center currentPosition is now 0
3/2/2021, 9:47:56 PM [HOOBS 21EA@Living Room Dining@Living Room Dining Battery@Battery Level] characteristic was supplied illegal value: number -1 exceeded minimum of 0
3/2/2021, 9:47:56 PM [MySmartBlindsBridge] Living Room Dining currentPosition is now 0
3/2/2021, 9:47:56 PM [HOOBS 21EA@Living Room Right@Living Room Right Battery@Battery Level] characteristic was supplied illegal value: number -1 exceeded minimum of 0
3/2/2021, 9:47:56 PM [MySmartBlindsBridge] Living Room Right currentPosition is now 0

**Comment by apexad** [comment]
I'll look into this soon, been taking a bit of a break from working on homebridge plugins. Hope to get back into it soon.

As far as Node version, there should generally be no issue with using 12.x instead of 14.x but as a note of clarity I am personally running Homebridge Config UI-X, Homebridge 1.3.0 and Node JS 14.x when I tests this plugin.

**Comment by randymassey** [comment]
Hi Alex, interesting, yes from what I read in the documentation there should be no issues in version 12.x.x, it’s mostly complaining randomly in the logging & as I recall only initially, eg on boot. 

But that is interesting about your configuration, because looking at the other posters log in this email, I’ve also been working with Kevin the author of the WizLAN plug-in, & we have been seeing similar references with the slow to respond commenting in the logs, for that plug-in it relates to the LED bulbs obviously. But what I would say is I am seeing this only for the first time in the logs with both plug-ins now since upgrading HOOBS to 3.3.3.

So two plug-ins do seem to be having somewhat parallel complaints in the logging. Hmmmm.  I wonder if we need to loop in someone in  HOOBS support group as well?

I do pay attention to my logs pretty regularly a good couple of times a week and certainly anytime I do updates. 

So I’ve only begun seeing these types of log notices after upgrading today with the HOOBS 3.3.3 update.

As always with me, no rush with regard to time, you know I like to get the info back up the chain so to speak.  

At  your leisure ….

I hope this finds you well.

Randy

**Comment by apexad** [comment]
So, I know homebridge 1.3.0 will now throw a lot more into the logs to 'bug' users to contact plugin authors for slow plugins and such.

Seems like HOOBS 3.3.3 is doing the same thing as well, but indeed they will need to verify that.  Certainly not a bad thing to be doing this as it makes troubleshooting your accessories a lot easier but there will likely be a need for plugin authors to pay attention to issues more often.

**Comment by randymassey** [comment]
Hi Alex, yes that makes sense, & in truth is logical as a progression, but it will begin to shift the hobble level tinker factor more to an actual development model…

I have debated abandoning HOOBS & using HB itself directly, but since I have 4 plugins working really well together & about 34 devices attached, it’s TRULY the last thing I actually WANT to do, hehehehe….  As I read, I hear tale of HB tending to operate more smoothly than HOOBS, but it seems it;s almost like a teeter totter, once they do the next UD, then there is more parity of the various small issues generally reported.  So, it seems HOOBS is on a slight time delay relative to HB itself.

The funny thing is, I actually also am using the Starling HUB separately, as I have numerous Nest products, & that little box has been so rock solid, I cannot even compare the two.  I never have to reboot it, & other than its firmware updates it has been so solid its cray.  So, I am not using a Nest plugin on HOOBS.  As I recall, Starling was initiated by one of the HB developers, specifically focusing on only Nest, both original Nest accounts as well as Google Nest accounts.

I honestly never even have to think about that box its actually been that reliable & consistent.  I know since it’s purpose built, it’s by no means an actual comparison relative to HOOBS/HB, but I do at times wish HOOBS could behave more like that Starling HUB, LOL…

Anyway, what you say makes total sense.

I always appreciate your input too BTW.  It’s great communicating with you sir…

Stay safe & well…

Randy

**Comment by rnewhart** [comment]
Circling back around to this.  I have 7 mysmartblinds and do you think controlling them all at once has anything to do with why some don't respond?   The bridge app never fails, but hoobs sending out individual requests to each blind seems to trip it up.

**Comment by apexad** [comment]
Yes, if I could modify this so it could send batch requests that would probably help. I’ll try to get to it soon, but don’t have much time currently.

**Comment by apexad** [comment]
Closing this in favor of #38 as the fix is the same, need to re-factor this plugin to do batch requests instead of single requests as the bridge is being flooded with requests and sometimes fails.

Hopefully will get to this soon.
